File: Matlab/utils/get_simpara.m
% ========================================================================
% Learning a Non-Locally Regularized Convolutional Sparse Representation
% for Joint Chromatic and Polarimetric Demosaicking, Version 1.0
% Copyright(c) 2024 Yidong Luo, Junchao Zhang, Jianbo Shao, Jiandong Tian,
% Jiayi Ma, All Rights Reserved.
%
% ----------------------------------------------------------------------
% Permission to use, copy, or modify this software and its documentation
% for educational and research purposes only and without fee is here
% granted, provided that this copyright notice and the original authors'
% names appear on all copies and supporting documentation. This program
% shall not be used, rewritten, or adapted as the basis of a commercial
% software or hardware product without first obtaining permission of the
% authors. The authors make no representations about the suitability of
% this software for any purpose. It is provided "as is" without express
% or implied warranty.
%
%----------------------------------------------------------------------
% This is an demo of "Learning a Non-Locally Regularized Convolutional 
% Sparse Representation for Joint Chromatic and Polarimetric Demosaicking."
%
% Please cite the following paper when you use it:
%
% Yidong Luo, Junchao Zhang, Jianbo Shao, Jiandong Tian, Jiayi Ma, "Learn- 
% -ing a Non-Locally Regularized Convolutional Sparse Representation for 
% Joint Chromatic and Polarimetric Demosaicking." IEEE Trans. on Image 
% Processing, 33, 5029-5044 (2024).
%----------------------------------------------------------------------

function B = get_simpara(S_pat, Sim_pat, Pat_arr, para)
% L2 parameters for NLSS constraint
f    = para.f;
ch   = para.ch;
L    = size(S_pat,4);
ns   = para.ns;
nSig = para.nSig;

if para.gpu_num > 0
    Cu0  = zeros(f,f,ch,L,'gpuArray');
    b0   = zeros(f,f,ch,L,'gpuArray');
else
    Cu0  = zeros(f,f,ch,L);
    b0   = zeros(f,f,ch,L);
end

for i = 1:L
    coe = S_pat(:,:,:,Pat_arr(:,i)) - repmat(Sim_pat(:,:,:,i),[1,1,1,ns]);
    Cu0(:,:,:,i) = mean(coe.^2,4);
    b0(:,:,:,i)  = S_pat(:,:,:,i) - Sim_pat(:,:,:,i);
end
Cu0 = max(0, Cu0-nSig^2);
b0  = (Cu0.*(b0.^2)).^(1/4);
B   = b0.^2 + para.Eps;
end

File: Python/utils/get_simpara.py
import numpy as np

try:
    import cupy as cp
    _HAS_CUPY = True
except Exception:
    cp = None
    _HAS_CUPY = False


def get_simpara(s_pat, sim_pat, pat_arr, para):
    f = para.f
    ch = para.ch
    l = s_pat.shape[3]
    ns = para.ns
    nsig = para.nSig

    use_cuda = _HAS_CUPY and isinstance(s_pat, cp.ndarray)
    if use_cuda:
        neighbors = s_pat[:, :, :, pat_arr]
        coe = neighbors - sim_pat[:, :, :, None, :]
        cu0 = cp.mean(coe ** 2, axis=3)
        b0 = s_pat - sim_pat
        cu0 = cp.maximum(0.0, cu0 - nsig ** 2)
        b0 = (cu0 * (b0 ** 2)) ** 0.25
        b = b0 ** 2 + para.Eps
        return b

    cu0 = np.zeros((f, f, ch, l), dtype=np.float64)
    b0 = np.zeros((f, f, ch, l), dtype=np.float64)

    for i in range(l):
        coe = s_pat[:, :, :, pat_arr[:, i]] - np.tile(sim_pat[:, :, :, i:i + 1], (1, 1, 1, ns))
        cu0[:, :, :, i] = np.mean(coe ** 2, axis=3)
        b0[:, :, :, i] = s_pat[:, :, :, i] - sim_pat[:, :, :, i]

    cu0 = np.maximum(0.0, cu0 - nsig ** 2)
    b0 = (cu0 * (b0 ** 2)) ** 0.25
    b = b0 ** 2 + para.Eps

    return b
